git-branchless/src/commands/move.rs

```diff
@@ -169,7 +170,57 @@ pub fn r#move(
     let repo_pool = RepoResource::new_pool(&repo)?;
     let rebase_plan = {
         let mut builder = RebasePlanBuilder::new(&dag);
+
         builder.move_subtree(source_oid, dest_oid)?;
+
+        if insert {
+            let source_head = {
+                let source_heads: CommitSet = dag
+                    .query()
+                    .heads(dag.query().descendants(CommitSet::from(source_oid))?)?;
+                match commit_set_to_vec_unsorted(&source_heads)?[..] {
+                    [oid] => oid,
+                    _ => {
+                        writeln!(
+                            effects.get_output_stream(),
+                            "The --insert flag cannot be used when moving subtrees with multiple heads."
+                        )?;
+                        return Ok(ExitCode(1));
+                    }
+                }
+            };
+
+            let dest_children: CommitSet = dag
+                .query()
+                .children(CommitSet::from(dest_oid))?
+                .difference(&dag.obsolete_commits);
+
+            for dest_child in commit_set_to_vec_unsorted(&dest_children)? {
+                if dag
+                    .query()
+                    .descendants(CommitSet::from(dest_child))?
+                    .contains(&source_oid.into())?
+                {
+                    // If this child subtree actually contains the source
+                    // subtree being moved, then we should only move the commit
+                    // range *up to* the source subtree, not the entire child
+                    // subtree.
```

Comment:
[no action required] I think there's an argument to be made that this logic ought to be in the constraint solver itself, i.e. transform any `move_subtree` constraints into `move_range` constraints if they would overlap without another `move_subtree` constraint. That being said, it sounds hard, and isn't necessary right now.